File: genium/src/main/java/com/here/genium/model/swift/SwiftType.java
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.model.swift;

public class SwiftType extends SwiftModelElement {

  public static final SwiftType INT8 = new SwiftType("Int8", "int8_t");
  public static final SwiftType UINT8 = new SwiftType("UInt8", "uint8_t");
  public static final SwiftType INT16 = new SwiftType("Int16", "int16_t");
  public static final SwiftType UINT16 = new SwiftType("UInt16", "uint16_t");
  public static final SwiftType INT32 = new SwiftType("Int32", "int32_t");
  public static final SwiftType UINT32 = new SwiftType("UInt32", "uint32_t");
  public static final SwiftType INT64 = new SwiftType("Int64", "int64_t");
  public static final SwiftType UINT64 = new SwiftType("UInt64", "uint64_t");

  public enum TypeCategory {
    BUILTIN_SIMPLE,
    BUILTIN_STRING,
    BUILTIN_BYTEBUFFER,
    STRUCT,
    ENUM,
    CLASS,
    ARRAY,
    DICTIONARY
  }

  public static final SwiftType VOID = new SwiftType("Void", null);
  public static final SwiftType BOOL = new SwiftType("Bool", "bool");
  public static final SwiftType FLOAT = new SwiftType("Float", "float");
  public static final SwiftType DOUBLE = new SwiftType("Double", "double");
  public static final SwiftType STRING =
      new SwiftType("String", "std_string", TypeCategory.BUILTIN_STRING);
  public static final SwiftType DATA =
      new SwiftType("Data", "data", TypeCategory.BUILTIN_BYTEBUFFER);

  public final String cPrefix;
  public final TypeCategory category;
  public final String implementingClass;
  public final String publicName;
  public final String className;
  public final boolean optional;

  public SwiftType(final String name, final String cPrefix) {
    this(name, cPrefix, TypeCategory.BUILTIN_SIMPLE);
  }

  public SwiftType(final String name, final String cPrefix, final TypeCategory category) {
    this(name, cPrefix, null, category, null, name, false);
  }

  protected SwiftType(
      final String name,
      final String cPrefix,
      final SwiftVisibility visibility,
      final TypeCategory category,
      final String implementingClass,
      final String publicName,
      final boolean optional) {
    super(name, visibility);
    this.cPrefix = cPrefix;
    this.optional = optional;
    this.category = category;
    this.implementingClass = implementingClass;
    this.publicName = publicName;
    this.className = category == TypeCategory.CLASS ? publicName : "";
  }

  public SwiftType withAlias(final String aliasName) {
    return new SwiftType(
        name, cPrefix, visibility, category, implementingClass, aliasName, optional);
  }

  public SwiftType withOptional(final boolean isOptional) {
    return this.optional != isOptional
        ? new SwiftType(
            name, cPrefix, visibility, category, implementingClass, publicName, isOptional)
        : this;
  }
}

File: gluecodium/src/main/java/com/here/gluecodium/model/swift/SwiftType.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.gluecodium.model.swift

open class SwiftType protected constructor(
    name: String,
    val cPrefix: String = "",
    visibility: SwiftVisibility? = null,
    val category: TypeCategory,
    val publicName: String = name,
    val optional: Boolean = false,
    val externalFramework: String? = null
) : SwiftModelElement(name, visibility) {
    val className
        get() = if (category == TypeCategory.CLASS) mangledName else ""

    open val mangledName
        get() = name.replace("_", "_1").replace(".", "_") + "_"

    constructor(
        name: String,
        cPrefix: String = "",
        category: TypeCategory = TypeCategory.BUILTIN_SIMPLE
    ) : this(name, cPrefix, null, category, name, false)

    enum class TypeCategory {
        BUILTIN_SIMPLE,
        BUILTIN_STRING,
        BUILTIN_OTHER,
        STRUCT,
        ENUM,
        CLASS,
        ARRAY,
        DICTIONARY,
        SET,
        CLOSURE,
        EXTENSION
    }

    // Has to be a function. For a property Kotlin will generate a getter with "C" capitalized.
    @Suppress("unused")
    fun getcPrefix() = cPrefix

    @Suppress("unused")
    val isExternal
        get() = externalFramework != null

    open fun withAlias(aliasName: String) =
        SwiftType(name, cPrefix, visibility, category, aliasName, optional, externalFramework)

    open fun withOptional(optional: Boolean) =
        when (optional) {
            this.optional -> this
            else -> SwiftType(name, cPrefix, visibility, category, publicName, optional, externalFramework)
        }

    companion object {
        val INT8 = SwiftType("Int8", "int8_t")
        val UINT8 = SwiftType("UInt8", "uint8_t")
        val INT16 = SwiftType("Int16", "int16_t")
        val UINT16 = SwiftType("UInt16", "uint16_t")
        val INT32 = SwiftType("Int32", "int32_t")
        val UINT32 = SwiftType("UInt32", "uint32_t")
        val INT64 = SwiftType("Int64", "int64_t")
        val UINT64 = SwiftType("UInt64", "uint64_t")

        val VOID = SwiftType("Void")
        val BOOL = SwiftType("Bool", "bool")
        val FLOAT = SwiftType("Float", "float")
        val DOUBLE = SwiftType("Double", "double")
        val STRING = SwiftType("String", "std_string", TypeCategory.BUILTIN_STRING)
        val DATA = SwiftType("Data", "data", TypeCategory.BUILTIN_OTHER)
        val DATE = SwiftType("Date", "date", TypeCategory.BUILTIN_OTHER)
        val LOCALE = SwiftType("Locale", "locale", TypeCategory.BUILTIN_OTHER)
    }
}
